File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> pq = new PriorityQueue<>((l1, l2) -> l1.val - l2.val);
        
        if (lists.length == 0)
            return null;
        
        for (int i = 0; i < lists.length; i++) {
            ListNode temp = lists[i];
            while (temp != null) {
                pq.add(temp);
                temp = temp.next;
            }
        }
        
        ListNode head = pq.poll();
        ListNode ref = head;
        
        while (!pq.isEmpty()) {
            ref.next = pq.poll();
            ref = ref.next;
        }
        if (ref != null)
            ref.next = null;
        
        return head;
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        
        pq = []
        for head in lists:
            while head:
                heapq.heappush(pq, head.val)
                head = head.next
        
        if not pq:
            return None
        
        head = ListNode(heapq.heappop(pq))
        ref = head
        while pq:
            ref.next = ListNode(heapq.heappop(pq))
            ref = ref.next
            
        return head